File: api/cbpmfio_accessors.c
//
// cbpmfio_utils.c
//

#include "cbpmfio.h"


// Provide the string reflecting the measurement timestamp.
// This string has a maxmimum length of CBI_MAX_STRING_LENGTH.
int cbpm_rawfile_timestamp_(char *timestamp) {

  strcpy( timestamp, (*p_v3_file_header).timestamp );

  return CBPMFIO_SUCCESS;

}


// Provide the number of turns for the measurement found in the file.
int cbpm_rawfile_num_instruments(void) {
  return p_v3_file_header->num_instruments;
}


// Provide the number of bunches for the measurement found in the file.
int cbpm_rawfile_num_bunches(void) {
  return (*p_v3_file_header).num_bunches;
}


// Provide the number of turns for the measurement found in the file.
int cbpm_rawfile_num_turns(void) {
  return (*p_v3_file_header).num_turns;
}


// Returns the value of the CESR CONDX number obtained from the
// CESR databse when a data file is written.
int cbpm_rawfile_condx_number(void) {
  return p_v3_cesr_header->condx;
}


// Returns a floating point value representing the CESR CERN total
// current measurement stored when the BPM measurement was taken.
float cbpm_rawfile_cern_current(void) {
  return p_v3_cesr_header->mA_current;
}



// Provide a single instrument name/location identifier string from the
// given data file each time this function is called until there are no
// more left to retrieve in the file.  At that point, this function
// will place an empty string in name.
int cbpm_rawfile_inst_names_(char *name) {

  int iidx;

  for (iidx = inst_count_g; iidx <= p_v3_file_header->num_instruments; iidx++) {

    if (iidx == inst_count_g) {
    
      if (iidx == p_v3_file_header->num_instruments) {
	inst_count_g = 0;
	strcpy( name, "\0" );
	return CBPMFIO_FAILURE;
      } else {
	inst_count_g++;
	strcpy( name, p_v3_inst_array[iidx].location );
	return CBPMFIO_SUCCESS;
      }

    }
    
  } //endFOR
  return CBPMFIO_FAILURE;

}



// Returns the CESR RF bucket number for the bunch index provided.
// Automatically takes into account the mapping of the bunch indexes
// for the timing setup the instruments were in when the measurement
// was performed.
//
// bunch index argument starts at 1.
// RF buckets are indexed here starting from 1.
// If bunch index is out of range for the timing setup found, return -1.
int cbpm_rawfile_rf_bucket(int *bunch) {

  int retval = -1;

  // 4ns timing setups
  if (p_v3_file_header->timing_setup[0] == '4') {
    if (*bunch > 640) {
      retval = -1;
    }
    retval = *bunch * 2;
  }

  // 14ns timing setups
  if (p_v3_file_header->timing_setup[0] == '1') {
    if (*bunch > 183) {
      retval = -1;
    }
    retval = (((*bunch-1)*14) / 2) + 1;
  }

  return retval;

}



// Return pedestal value for a given instrument, bunch, and button.
float cbpm_rawfile_pedestal( char *inst_location, int *bunch, int *button ) {
  
  // Necessary for making a proper comparison between the string passed
  // in from Fortran (potentially with padding spaces) and the C string.
  cbpmfio_strip(inst_location);

  int iidx;
  int array_idx = (*bunch)-1;
  float retval = 0.0;
  int tblock;
  
  for (iidx = 0; iidx <  p_v3_file_header->num_instruments; iidx++) {
    if (strcmp( inst_location, p_v3_inst_array[iidx].location) == 0 ) { 
      
      tblock = p_v3_inst_array[iidx].timing_blocks[array_idx]; 	
      retval = p_v3_inst_array[iidx].pedestals[tblock][*button];
      return retval;
      
    } //endIF
  } //endFOR
  return CBPMFIO_FAILURE;
}


// Return gain scale factor for a given instrument, bunch, and button.
float cbpm_rawfile_gain_factor( char *inst_location, int *bunch, int *button ) {

  // Necessary for making a proper comparison between the string passed
  // in from Fortran (potentially with padding spaces) and the C string.
  cbpmfio_strip(inst_location);

  int iidx, tblock;
  int array_idx = (*bunch)-1;
  float retval;
    
  for (iidx = 0; iidx <  p_v3_file_header->num_instruments; iidx++) {
    if (strcmp( inst_location, p_v3_inst_array[iidx].location) == 0 ) { 

      tblock = p_v3_inst_array[iidx].timing_blocks[array_idx]; 	
      retval = p_v3_inst_array[iidx].gain_coeffs[tblock][*button];
      return retval;

    } //endIF
  } //endFOR
  return CBPMFIO_FAILURE;
}



// Provide a pointer to the raw ADC value buffers for a given instrument and button.
int *cbpm_rawfile_raw_ptr(char *inst_location, int *button) {

  // Necessary for making a proper comparison between the string passed
  // in from Fortran (potentially with padding spaces) and the C string.
  cbpmfio_strip(inst_location);

  int  iidx;

  for (iidx = 0; iidx <  p_v3_file_header->num_instruments; iidx++) {
    if (strcmp( inst_location, p_v3_inst_array[iidx].location) == 0 ) {

      switch (*button) {
      case 0:
	return (int *)p_v3_inst_array[iidx].TIraw;
	break;
      case 1:
	return (int *)p_v3_inst_array[iidx].BIraw;
	break;
      case 2:
	return (int *)p_v3_inst_array[iidx].BOraw;
	break;
      case 3:
	return (int *)p_v3_inst_array[iidx].TOraw;
	break;
      default:
	return NULL;
	break;
      }

    } //endIF strcmp

  } //endFOR iidx

  // Instrument not found.
  return NULL;

}



// Provide a pointer to the scaled data buffers for the given instrument and button.
float *cbpm_rawfile_tbt_ptr(char *inst_location, int *button) {

  // Necessary for making a proper comparison between the string passed
  // in from Fortran (potentially with padding spaces) and the C string.
  cbpmfio_strip(inst_location);

  int  iidx;

  for (iidx = 0; iidx <  p_v3_file_header->num_instruments; iidx++) {
    if (strcmp( inst_location, p_v3_inst_array[iidx].location) == 0 ) {

      switch (*button) {
      case 0:
	return (float *)p_v3_inst_array[iidx].TIcorrected;
	break;
      case 1:
	return (float *)p_v3_inst_array[iidx].BIcorrected;
	break;
      case 2:
	return (float *)p_v3_inst_array[iidx].BOcorrected;
	break;
      case 3:
	return (float *)p_v3_inst_array[iidx].TOcorrected;
	break;
      default:
	return NULL;
	break;
      }

    } //endIF strcmp

  } //endFOR iidx

  return NULL;

}


float cbpmfio_compute_pos(int plane, float coeffs[], float TI, float BI, float BO, float TO) {

  float Xcoeff, Ycoeff, Xnum, Ynum, Xpos, Ypos, sum;
  
  Xcoeff = coeffs[0];
  Ycoeff = coeffs[1];

  sum  = TI + BI + BO + TO;
  Xnum = TO + BO - BI - TI;
  Ynum = TO - BO - BI + TI;

  if (plane == 0) {
    Xpos = Xcoeff * (Xnum/sum);
    return Xpos;
  } else {
    Ypos = Ycoeff * (Ynum/sum);
    return Ypos;
  }


}


// Provide a pointer to the calculated position buffer for the given plane.
float* cbpm_rawfile_pos_ptr(char *inst_location, int *plane) {

  // Necessary for making a proper comparison between the string passed
  // in from Fortran (potentially with padding spaces) and the C string.
  cbpmfio_strip(inst_location);

  int  iidx;

  for (iidx = 0; iidx <  p_v3_file_header->num_instruments; iidx++) {
    if (strcmp( inst_location, p_v3_inst_array[iidx].location) == 0 ) {

      switch (*plane) {
      case 0:
	return (float *)p_v3_inst_array[iidx].Xpos;
	break;
      case 1:
	return (float *)p_v3_inst_array[iidx].Ypos;
	break;
      default:
	return NULL;
	break;
      }

    } //endIF strcmp

  } //endFOR iidx
  
  // Instrument not found.
  return NULL;

}


// Provide a pointer to the phase word buffers for the given instrument and button.
int *cbpm_rawfile_phase_ptr(char *inst_location, int *plane) {

  // Necessary for making a proper comparison between the string passed
  // in from Fortran (potentially with padding spaces) and the C string.
  cbpmfio_strip(inst_location);

  int  iidx;

  for (iidx = 0; iidx <  p_v3_file_header->num_instruments; iidx++) {
    if (strcmp( inst_location, p_v3_inst_array[iidx].location) == 0 ) {

      switch (*plane) {
      case 0:
	return (int *)p_v3_inst_array[iidx].X_phase_words;
	break;
      case 1:
	return (int *)p_v3_inst_array[iidx].Y_phase_words;
	break;
      default:
	return NULL;
	break;
      }

    } //endIF strcmp

  } //endFOR iidx
  
  // Instrument not found.
  return NULL;

}

File: api/cbpm_fio_accessors.f90
#include "cbpmfio_f_constants.h"

module cbpm_file_input

use,intrinsic :: iso_c_binding

interface
   function cbpm_read_rawfile_f(file_idx) result (val) bind(c,name='cbpm_read_rawfile')
     Integer file_idx
     Integer val
   end function cbpm_read_rawfile_f
end interface

interface
   function cbpm_rawfile_num_instruments_f() result (val) bind(c,name='cbpm_rawfile_num_instruments')
     Integer val
   end function cbpm_rawfile_num_instruments_f
end interface

interface
   function cbpm_rawfile_num_bunches_f() result (val) bind(c,name='cbpm_rawfile_num_bunches')
     Integer val
   end function cbpm_rawfile_num_bunches_f
end interface

interface
   function cbpm_rawfile_num_turns_f() result (val) bind(c,name='cbpm_rawfile_num_turns')
     Integer val
   end function cbpm_rawfile_num_turns_f
end interface

interface
   function cbpm_rawfile_condx_number_f() result (val) bind(c,name='cbpm_rawfile_condx_number')
     Integer val
   end function cbpm_rawfile_condx_number_f
end interface

interface
   function cbpm_rawfile_cern_current_f() result (val) bind(c,name='cbpm_rawfile_cern_current')
     Integer val
   end function cbpm_rawfile_cern_current_f
end interface



interface
   function cbpm_rawfile_rf_bucket_f(bunch) result (val) bind(c,name='cbpm_rawfile_rf_bucket')
     Integer bunch
     Integer val
   end function cbpm_rawfile_rf_bucket_f
end interface




contains


!--------------------------
  Function cbpm_rawfile_timestamp_f( timestamp ) result (val)

  Implicit None

  Integer val
  Integer i
  Integer, external ::  cbpm_rawfile_timestamp

  Character(*) timestamp
  Character(CBPMFIO_MAX_STRING_LENGTH) temptimestamp

  val = cbpm_rawfile_timestamp( temptimestamp )


  ! Strip off the null character before passing the string back.
  do i = 1, CBPMFIO_MAX_STRING_LENGTH
     if ( ichar(temptimestamp(i:i)) == 0) then
        timestamp = temptimestamp(1:i-1)
        exit
     end if
  end do

end Function cbpm_rawfile_timestamp_f



Function cbpm_rawfile_inst_names_f( name ) result (val)
  
  Implicit None
  
  Integer val
  Integer ccount
  Integer, external :: cbpm_rawfile_inst_names
  
  Character(*) name
  Character(CBPMFIO_MAX_STRING_LENGTH) tempname

  val = cbpm_rawfile_inst_names( tempname )

  ! Strip off the null character before passing the string back.
  do ccount = 1, CBPMFIO_MAX_STRING_LENGTH
     if ( ichar(tempname(ccount:ccount)) == 0) then
        name = tempname(1:ccount-1)
        exit
     end if
  end do
  
end Function cbpm_rawfile_inst_names_f



!!--------------------------------------------------------------------------
!! Intermediate layer definitions to access data buffers follow
!!--------------------------------------------------------------------------

function cbpm_rawfile_pedestal_f(inst_location, bunch, button) result (val)

  Implicit None
  character(*) inst_location
  Integer bunch, button
  Real val

  interface
     function cbpm_rawfile_pedestal(inst_location, bunch, button) result (val) bind(c)
       import c_char
       Character(c_char) inst_location(*)
       Integer bunch, button
       Real val
     end function cbpm_rawfile_pedestal
  end interface
  
  val = cbpm_rawfile_pedestal(trim(inst_location)//C_NULL_CHAR, bunch, button)
  
end function cbpm_rawfile_pedestal_f




function cbpm_rawfile_gain_factor_f(inst_location, bunch, button) result (val)

  Implicit None
  character(*) inst_location
  Integer bunch, button
  Real val

  interface
     function cbpm_rawfile_gain_factor(inst_location, bunch, button) result (val) bind(c)
       import c_char
       Character(c_char) inst_location(*)
       Integer bunch, button
       Real val
     end function cbpm_rawfile_gain_factor
  end interface
  
  val = cbpm_rawfile_gain_factor(trim(inst_location)//C_NULL_CHAR, bunch, button)
  
end function cbpm_rawfile_gain_factor_f
  



function cbpm_rawfile_raw_ptr_f(inst_location, button) result (dptr)
  
  Implicit None
  Integer, Pointer :: dptr(:)
  Type (c_ptr) c_pointer
  character(*) inst_location
  Integer button, num_bunches, num_turns

  interface
     function cbpm_rawfile_raw_ptr(inst_location, button) result (c_pointer) bind(c)
       import c_ptr, c_char
       Character(c_char) inst_location(*)
       Integer button
       type (c_ptr) :: c_pointer
     end function cbpm_rawfile_raw_ptr
  end interface

  c_pointer = cbpm_rawfile_raw_ptr(trim(inst_location)//C_NULL_CHAR, button)
  num_bunches = cbpm_rawfile_num_bunches_f()
  num_turns = cbpm_rawfile_num_turns_f()
  call c_f_pointer(c_pointer, dptr, [num_bunches*num_turns])

end function cbpm_rawfile_raw_ptr_f




function cbpm_rawfile_tbt_ptr_f(inst_location, button) result (dptr)

  Implicit None
  Real*4, Pointer :: dptr(:)
  Type (c_ptr) c_pointer
  character(*) inst_location
  Integer button, num_bunches, num_turns

  interface
     function cbpm_rawfile_tbt_ptr(inst_location, button) result (c_pointer) bind(c)
       import c_ptr, c_char
       Character(c_char) inst_location(*)
       Integer button
       type (c_ptr) :: c_pointer
     end function cbpm_rawfile_tbt_ptr
  end interface
  
  c_pointer = cbpm_rawfile_tbt_ptr(trim(inst_location)//C_NULL_CHAR, button)
  num_bunches = cbpm_rawfile_num_bunches_f()
  num_turns = cbpm_rawfile_num_turns_f()

  call c_f_pointer(c_pointer, dptr, [num_bunches*num_turns])

end function cbpm_rawfile_tbt_ptr_f




function cbpm_rawfile_pos_ptr_f(inst_location, plane) result (dptr)

  Implicit None
  Real*4, Pointer :: dptr(:)
  Type (c_ptr) c_pointer
  character(*) inst_location
  Integer plane, num_bunches, num_turns

  interface
     function cbpm_rawfile_pos_ptr(inst_location, plane) result (c_pointer) bind(c)
       import c_ptr, c_char
       Character(c_char) inst_location(*)
       Integer plane
       type (c_ptr) :: c_pointer
     end function cbpm_rawfile_pos_ptr
  end interface

  c_pointer = cbpm_rawfile_pos_ptr(trim(inst_location)//C_NULL_CHAR, plane)
  num_bunches = cbpm_rawfile_num_bunches_f()
  num_turns = cbpm_rawfile_num_turns_f()
  call c_f_pointer(c_pointer, dptr, [num_bunches*num_turns])

end function cbpm_rawfile_pos_ptr_f




function cbpm_rawfile_phase_ptr_f(inst_location, plane) result (dptr)
  
  Implicit None
  Integer, Pointer :: dptr(:)
  Type (c_ptr) c_pointer
  character(*) inst_location
  Integer plane, num_turns

  interface
     function cbpm_rawfile_phase_ptr(inst_location, plane) result (c_pointer) bind(c)
       import c_ptr, c_char
       Character(c_char) inst_location(*)
       Integer plane
       type (c_ptr) :: c_pointer
     end function cbpm_rawfile_phase_ptr
  end interface

  c_pointer = cbpm_rawfile_phase_ptr(trim(inst_location)//C_NULL_CHAR, plane)
  num_turns = cbpm_rawfile_num_turns_f()
  call c_f_pointer(c_pointer, dptr, [num_turns])

end function cbpm_rawfile_phase_ptr_f



end module cbpm_file_input
